helper/decrypt_check.go

```diff
@@ -0,0 +1,101 @@
+package helper
+
+import (
+	"bytes"
+	"crypto/aes"
+	"crypto/cipher"
+	"io"
+
+	"github.com/ProtonMail/go-crypto/openpgp/packet"
+	"github.com/ProtonMail/gopenpgp/v2/crypto"
+	"github.com/pkg/errors"
+)
+
+const AES_BLOCK_SIZE = 16
+
+func supported(cipher packet.CipherFunction) bool {
+	switch cipher {
+	case packet.CipherAES128, packet.CipherAES192, packet.CipherAES256:
+		return true
+	case packet.CipherCAST5, packet.Cipher3DES:
+		return false
+	}
+	return false
+}
+
+func blockSize(cipher packet.CipherFunction) int {
+	switch cipher {
+	case packet.CipherAES128, packet.CipherAES192, packet.CipherAES256:
+		return AES_BLOCK_SIZE
+	case packet.CipherCAST5, packet.Cipher3DES:
+		return 0
+	}
+	return 0
+}
+
+func blockCipher(cipher packet.CipherFunction, key []byte) (cipher.Block, error) {
+	switch cipher {
+	case packet.CipherAES128, packet.CipherAES192, packet.CipherAES256:
+		return aes.NewCipher(key)
+	case packet.CipherCAST5, packet.Cipher3DES:
+		return nil, errors.New("gopenpgp: cipher not supported for quick check")
+	}
+	return nil, errors.New("gopenpgp: unknown cipher")
+}
+
+// QuickCheckDecryptReader checks with high probability if the provided session key
+// can decrypt a data packet given its 24 byte long prefix.
+// The method reads up to but not exactly 24 bytes from the prefixReader.
+// NOTE: Only works for SEIPDv1 packets with AES.
+func QuickCheckDecryptReader(sessionKey *crypto.SessionKey, prefixReader crypto.Reader) (bool, error) {
+	algo, err := sessionKey.GetCipherFunc()
+	if err != nil {
+		return false, errors.New("gopenpgp: cipher algorithm not found")
+	}
+	if !supported(algo) {
+		return false, errors.New("gopenpgp: cipher not supported for quick check")
+	}
+	packetParser := packet.NewReader(prefixReader)
+	_, err = packetParser.Next()
+	if err != nil {
+		return false, errors.New("gopenpgp: failed to parse packet prefix")
+	}
+
+	blockSize := blockSize(algo)
+	encryptedData := make([]byte, blockSize+2)
+	_, err = io.ReadFull(prefixReader, encryptedData)
+	if err != nil {
+		return false, errors.New("gopenpgp: prefix is too short to check")
+	}
+
+	blockCipher, err := blockCipher(algo, sessionKey.Key)
+	if err != nil {
+		return false, errors.New("gopenpgp: failed to initialize the cipher")
+	}
+
+	blockBuffer := make([]byte, blockSize)
+	// Decrypt 2 bytes of the second block
+	blockCipher.Encrypt(blockBuffer, encryptedData[:blockSize])
+	for ind := range blockBuffer[:2] {
+		encryptedData[blockSize+ind] ^= blockBuffer[ind]
+	}
+	for ind := range blockBuffer {
+		blockBuffer[ind] = 0
+	}
+	// Decrypt the first block
+	blockCipher.Encrypt(blockBuffer, blockBuffer)
+	for ind := range blockBuffer {
+		encryptedData[ind] ^= blockBuffer[ind]
+	}
```

Comment:
Just to note: it might be simpler to create a CFB instance, which would do this for you? But, not critical, we can merge it as-is and change it later

Reply:
Good point, I agree. It is slightly more expensive (1 more buffer and 2 copies), but should not be noticeable and there is no code duplication. 

[f1e673a](https://github.com/ProtonMail/gopenpgp/pull/249/commits/f1e673ab63f95fa3c181c9b1b419e76d40b1a164) integrates the change.